### PR TITLE
feat(dataset): implement Mapping-related methods in class "Dataset"

### PR DESCRIPTION
**Pull request by graczhual**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2ccta6r 实现 Dataset 类中的 Mapping 相关方法，支持用户读取 sheets](https://app.clickup.com/t/2ccta6r)

File: graviti/dataset/dataset.py [hunk 1]
```diff
@@ -5,20 +5,35 @@
...
 from graviti.dataframe.frame import DataFrame
 from graviti.manager.branch import BranchManager
 from graviti.manager.commit import CommitManager
 from graviti.manager.draft import DraftManager
 from graviti.manager.tag import TagManager
+from graviti.utility.lazy import LazyFactory
 
 
-class Dataset(UserMutableMapping[str, DataFrame]):
+def _getter(func: Callable[..., Any]) -> Callable[..., Any]:
```
Comment (linjiX) [inline graviti/dataset/dataset.py:25]: decorator should use TypeVar to indicate the input type is the same as output type.
ref: https://github.com/linjiX/tensorbay-python-sdk/blob/e03ae156c9db410528146b26054796171d54d1f8/tensorbay/utility/deprecated.py#L45

File: graviti/dataset/dataset.py [hunk 2]
```diff
@@ -49,6 +64,31 @@ def __init__(
         self._alias = alias
         self._is_public: Optional[bool] = None
 
+        self._is_loaded: bool = False
```
Comment (linjiX) [inline graviti/dataset/dataset.py:67]: Is the type hint reduntant?

Comment (linjiX) [reply graviti/dataset/dataset.py:67]: Do we need `_is_loaded`?
can we just use `_data` as a flag?

File: graviti/dataset/dataset.py [hunk 3]
```diff
@@ -196,5 +236,53 @@ def upload(
...
+            sheet_name = sheet["name"]
+            data_details = list_data_details(
+                self._url,
+                self._access_key,
+                self._dataset_id,
+                sheet_name,
+                draft_number=self._status.draft_number,
+                commit=self._status.commit_id,
+            )
+            factory_getter = partial(
```
Comment (linjiX) [inline graviti/dataset/dataset.py:258]: why use `lambda` with `partial`??????????

File: graviti/dataframe/frame.py [hunk 4]
```diff
@@ -75,6 +78,18 @@ def __repr__(self) -> str:
     def __len__(self) -> int:
         pass
 
+    @classmethod
+    def from_dataset(cls: Type[_T], factory: LazyFactory) -> _T:
```
Comment (linjiX) [inline graviti/dataframe/frame.py:82]: the method name `from_dataset` is not accurate, because the input is a factory

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataset/dataset.py [hunk 5]
```diff
@@ -5,20 +5,37 @@
...
+from graviti.utility.lazy import LazyFactory
 
+_Callable = TypeVar("_Callable", bound=Callable[..., Any])
 
-class Dataset(UserMutableMapping[str, DataFrame]):
+
+def _getter(func: _Callable) -> _Callable:
+    @wraps(func)
+    def wrapper(self: "Dataset", *args: Any, **kwargs: Any) -> Any:
+        if not hasattr(self, "_is_loaded"):
```
Comment (linjiX) [inline graviti/dataset/dataset.py:30]: `_is_loaded` is not defined?

File: graviti/dataset/dataset.py [hunk 6]
```diff
@@ -196,5 +237,54 @@ def upload(
...
+                    offset=offset,
+                    limit=limit,
+                )
+
+            factory = LazyFactory(
+                data_details["totalCount"],
+                128,
+                factory_getter,
+            )
+            # catalog = get_catalog(
```
Comment (linjiX) [inline graviti/dataset/dataset.py:280]: remove the useless comments

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataset/dataset.py (hunk 6 above)
Comment (graczhual) [reply graviti/dataset/dataset.py:280]: this will be used after `catalog_to_schema` lands

**Review by AChenQ — APPROVED** [review]

**Review by linjiX — APPROVED** [review]

File: graviti/dataset/dataset.py [hunk 7]
```diff
@@ -5,20 +5,37 @@
...
+_Callable = TypeVar("_Callable", bound=Callable[..., Any])
 
-class Dataset(UserMutableMapping[str, DataFrame]):
+
+def _getter(func: _Callable) -> _Callable:
+    @wraps(func)
+    def wrapper(self: "Dataset", *args: Any, **kwargs: Any) -> Any:
+        if not hasattr(self, "_data"):
+            self.load()
+        result = func(self, *args, **kwargs)
```
Comment (linjiX) [inline graviti/dataset/dataset.py:32]: `return func(self, *args, **kwargs)`

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/dataset/dataset.py [hunk 8]
```diff
@@ -49,6 +68,28 @@ def __init__(
...
+    @_getter
+    def __getitem__(self, key: str) -> DataFrame:
+        return self._data.__getitem__(key)
+
+    @_getter
+    def __iter__(self) -> Iterator[str]:
+        return self._data.__iter__()
+
+    @_getter
+    def keys(self) -> AbstractSet[str]:
```
Comment (linjiX) [inline graviti/dataset/dataset.py:84]: Should the return be `KeyView`?

File: graviti/dataset/dataset.py [hunk 9]
```diff
@@ -5,20 +5,37 @@
...
 from graviti.manager.commit import CommitManager
 from graviti.manager.draft import DraftManager
 from graviti.manager.tag import TagManager
+from graviti.utility.lazy import LazyFactory
 
+_Callable = TypeVar("_Callable", bound=Callable[..., Any])
 
-class Dataset(UserMutableMapping[str, DataFrame]):
+
+def _getter(func: _Callable) -> _Callable:
```
Comment (linjiX) [inline graviti/dataset/dataset.py:27]: Compare to decorator, will a method to check and return the `_data` be better?
ref: https://github.com/linjiX/tensorbay-python-sdk/blob/e03ae156c9db410528146b26054796171d54d1f8/tensorbay/client/lazy.py#L345

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Pull request merged**